### ISSUE TITLE
ChatHistory: Support forking messages (parent has n children)

### ISSUE THREAD

**Issue by dotnetprofessional** [microsoft/semantic-kernel#5834]
Support for forking of chats, which generates two or more messages from an original parent message. The ChatGpt UX will do this when performing a regeneration. See image for example (ie 2/2 is navigatable)

<img width="277" alt="image" src="https://github.com/microsoft/semantic-kernel/assets/5588330/9e6c7c0e-f78b-47c8-8fd9-c7ca9c295834">

The use case for this is when a user is brainstorming an idea. They may want to go back to a particular generation and then regenerate 1 or more times and take the path down again. Admittedly for this to be effective a good UX will need to support this too. However, I see a  lot of value in this scenario. This is the scenario I am building for today.

**Comment by dotnetprofessional** [comment]
Yes cloning is a solution but a messy one, as it does result in duplication of messages or at least their indexes. However, it also makes it harder to manage assuming you're building a similar UX to ChatGPT, as mentioned in the original post.

If SKs ChatHistory is only used as a vehicle to talk to the LLM then this can be resolved by maintaining a separate parent aware ChatHistory (effectively what I do now) then send the appropriate flattened history to SK. However, I'm not sure how this would work when adding persistence or other features. Ideally, I would be able to use one model rather than mapping and having to provide bespoke implementations for other features, such as storage etc.

**Comment by stephentoub** [comment]
I'm not clear on the request then. Are you asking for ChatHistory to be a tree rather than a list?

**Comment by dotnetprofessional** [comment]
Not necessarily a tree structure as in nodes, but metadata on what the parent for the message was. To be able to create a given thread (path) you'd also need to track the last messageId (active message) so you'd be able to traverse from that messageId back up using its parentId. If a consumer had no use for this pattern, then they could use it as normal just ignoring the parentIds.

Having a look at the retail chatGPT model, they are returning this data as metadata of the message:

        "metadata": {
          "messageType": "next",
          "modelSlug": "text-davinci-002-render-sha",
          "parentId": "05add3ae-b6eb-4476-8b0b-2ee09380a8f9"
        },

So I guess if we had the option to add custom metadata as key value pairs that was also persisted (ie we don't lose this data) it would achieve the goal and make it less specific and more extensible for other future use cases. The active messageId (ie the last message user is working with), seems like a UX concern so wouldn't need to be added either.
